**Bab 1: Pengenalan dan Pembuatan Array NumPy**

*1.1. Apa itu NumPy dan Keunggulannya?*

**NumPy** (singkatan dari Numerical Python) adalah library fundamental di Python untuk komputasi saintifik. Objek utamanya adalah ndarray (n-dimensional array), yaitu sebuah tabel data yang semua isinya harus memiliki **tipe data yang sama**.

Bayangkan sebuah List di Python bisa diisi berbagai macam data (angka, tulisan, dll). Nah, NumPy Array lebih spesifik, seperti sebuah kotak yang hanya bisa diisi oleh angka saja. Hal ini memberikan beberapa keuntungan besar dibanding List biasa:

- **Kecepatan**: Operasi matematika di NumPy jauh lebih cepat karena dieksekusi menggunakan bahasa C yang lebih efisien.

- **Memori**: NumPy Array membutuhkan ruang memori yang lebih sedikit daripada List Python untuk jumlah data yang sama.

- **Kenyamanan**: Memungkinkan kita melakukan operasi matematika pada seluruh data sekaligus tanpa perlu menulis perulangan (loop). Konsep ini disebut Vectorization.

*1.2. Membuat Array 1 Dimensi dari List*

In [1]:
# Pertama, kita import library NumPy dengan alias 'np' (ini adalah standar umum)
import numpy as np

# Sebuah list Python biasa
list_python = [1, 2, 3, 4, 5]

# Mengubah list menjadi array NumPy
array_numpy = np.array(list_python)

print(f"Ini adalah List Python: {list_python}")
print(f"Ini adalah Array NumPy: {array_numpy}")
print(f"Tipe data list_python: {type(list_python)}")
print(f"Tipe data array_numpy: {type(array_numpy)}")

Ini adalah List Python: [1, 2, 3, 4, 5]
Ini adalah Array NumPy: [1 2 3 4 5]
Tipe data list_python: <class 'list'>
Tipe data array_numpy: <class 'numpy.ndarray'>


*1.3. Membuat Array dengan Fungsi Bawaan*

In [2]:
import numpy as np

# Membuat array berisi lima angka 0
array_nol = np.zeros(5)
print(f"Array berisi angka nol: {array_nol}")

# Membuat array berisi empat angka 1
array_satu = np.ones(4)
print(f"Array berisi angka satu: {array_satu}")

# Membuat array dengan rentang nilai (dari 0 hingga sebelum 10, dengan langkah 2)
# Mirip seperti fungsi range() di Python
array_rentang = np.arange(0, 10, 2)
print(f"Array dengan rentang nilai: {array_rentang}")

Array berisi angka nol: [0. 0. 0. 0. 0.]
Array berisi angka satu: [1. 1. 1. 1.]
Array dengan rentang nilai: [0 2 4 6 8]


**Bab 2: Operasi Matematika (Vectorization)**

*2.1. Operasi Aritmatika Dasar*

In [3]:
import numpy as np

array_a = np.array([10, 20, 30, 40])
array_b = np.array([2, 4, 6, 8])

# Penjumlahan
print(f"Penjumlahan: {array_a + array_b}")

# Pengurangan
print(f"Pengurangan: {array_a - array_b}")

# Perkalian
print(f"Perkalian: {array_a * array_b}")

# Pembagian
print(f"Pembagian: {array_a / array_b}")

# Operasi dengan scalar
print(f"Dikali dua: {array_a * 2}")

Penjumlahan: [12 24 36 48]
Pengurangan: [ 8 16 24 32]
Perkalian: [ 20  80 180 320]
Pembagian: [5. 5. 5. 5.]
Dikali dua: [20 40 60 80]


*2.2. Fungsi Matematika Universal (UFuncs)*

In [4]:
import numpy as np

array_c = np.array([1, 4, 9, 16])

# Menghitung akar kuadrat (square root) setiap elemen
print(f"Akar kuadrat: {np.sqrt(array_c)}")

# Menghitung eksponensial (e^x) setiap elemen
print(f"Eksponensial: {np.exp(array_c)}")

# Menghitung logaritma natural setiap elemen
print(f"Logaritma: {np.log(array_c)}")

Akar kuadrat: [1. 2. 3. 4.]
Eksponensial: [2.71828183e+00 5.45981500e+01 8.10308393e+03 8.88611052e+06]
Logaritma: [0.         1.38629436 2.19722458 2.77258872]


*2.3 Operasi Vektor & Matriks*

In [5]:
#Operasi Vektor
import numpy as np
X = np.array([2.0,3.0,1.5])
Y = np.array([0.5,-1.0,0.8])

print("addition")
Z=X+Y
print(Z)
print("substraction")
Z=X-Y
print(Z)
print("Multiplication")
Z=X*Y
print(Z)
print("dot product")
Z=X@Y
print(Z)
print("dot product")
Z=np.dot(X,Y)
print(Z)

addition
[2.5 2.  2.3]
substraction
[1.5 4.  0.7]
Multiplication
[ 1.  -3.   1.2]
dot product
-0.7999999999999998
dot product
-0.7999999999999998


In [13]:
#Operasi Matriks
import numpy as np
X = np.array([[1, 2, 3],
              [0, 1, 4],
              [5, 6, 0]])
Y = np.array([[1,2,3],
              [4,5,6],
              [7,8,9]])

print("addition")
Z_add=X+Y
print(Z_add)
print("substraction")
Z_subs=X-Y
print(Z_subs)
print("Multiplication")
Z_multi=X*Y
print(Z_multi)
print("dot product")
Z_dot_1=X@Y
print(Z_dot_1)
print("dot product")
Z_dot_2=np.dot(X,Y)
print(Z_dot_2)
print("Matriks Transpose")
Z_transpose=X.T
print(Z_transpose)
print("Matriks Inverse")
Z_inverse=np.linalg.inv(X)
print(Z_inverse)
print("Matriks Identitas")
I=X@X_inverse
print(I)

addition
[[ 2  4  6]
 [ 4  6 10]
 [12 14  9]]
substraction
[[ 0  0  0]
 [-4 -4 -2]
 [-2 -2 -9]]
Multiplication
[[ 1  4  9]
 [ 0  5 24]
 [35 48  0]]
dot product
[[30 36 42]
 [32 37 42]
 [29 40 51]]
dot product
[[30 36 42]
 [32 37 42]
 [29 40 51]]
Matriks Transpose
[[1 0 5]
 [2 1 6]
 [3 4 0]]
Matriks Inverse
[[-24.  18.   5.]
 [ 20. -15.  -4.]
 [ -5.   4.   1.]]
Matriks Identitas
[[ 1.00000000e+00 -2.66453526e-15  0.00000000e+00]
 [ 0.00000000e+00  1.00000000e+00  0.00000000e+00]
 [ 0.00000000e+00 -7.10542736e-15  1.00000000e+00]]


**Bab 3: Indexing dan Slicing**

*3.1. Mengakses Elemen (Indexing)*

In [5]:
import numpy as np

# Array 1 Dimensi
arr_1d = np.array([10, 11, 12, 13, 14])
print(f"Elemen ke-2 dari arr_1d: {arr_1d[1]}") # Mengambil elemen kedua

# Array 2 Dimensi (matriks)
arr_2d = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
# Formatnya adalah array[baris, kolom]
print(f"Elemen pada baris 1, kolom 2 dari arr_2d: {arr_2d[0, 1]}") # Mengambil angka 2

Elemen ke-2 dari arr_1d: 11
Elemen pada baris 1, kolom 2 dari arr_2d: 2


*3.2. Mengambil Sebagian Elemen (Slicing)*

In [6]:
import numpy as np

arr = np.arange(10) # Membuat array [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

# Mengambil elemen dari indeks 2 hingga sebelum indeks 5
print(f"Slicing 1D: {arr[2:5]}")

# Mengambil sebagian dari array 2D
arr_2d = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])

# Mengambil 2 baris pertama dan 2 kolom pertama
# Baris: 0 hingga sebelum 2. Kolom: 0 hingga sebelum 2.
print("Slicing 2D:\n", arr_2d[0:2, 0:2])

Slicing 1D: [2 3 4]
Slicing 2D:
 [[1 2]
 [4 5]]


*3.3. Boolean Indexing*

In [7]:
import numpy as np

data_nilai = np.array([50, 85, 92, 65, 78, 95])

# Membuat kondisi boolean
lulus = data_nilai > 75
print(f"Kondisi boolean (lulus > 75): {lulus}")

# Menggunakan kondisi tersebut untuk menyaring array asli
print(f"Nilai yang lulus: {data_nilai[lulus]}")

Kondisi boolean (lulus > 75): [False  True  True False  True  True]
Nilai yang lulus: [85 92 78 95]


**Bab 4: Atribut dan Manipulasi Bentuk Array**

*4.1. Mengetahui Atribut Array*

In [8]:
import numpy as np

arr_2d = np.array([[1, 2, 3, 4], [5, 6, 7, 8]])

# .ndim: Mengetahui jumlah dimensi
print(f"Jumlah dimensi: {arr_2d.ndim}")

# .shape: Mengetahui bentuk array (jumlah baris, jumlah kolom)
print(f"Bentuk array: {arr_2d.shape}")

# .size: Mengetahui total elemen dalam array
print(f"Total elemen: {arr_2d.size}")

# .dtype: Mengetahui tipe data elemen di dalam array
print(f"Tipe data elemen: {arr_2d.dtype}")

Jumlah dimensi: 2
Bentuk array: (2, 4)
Total elemen: 8
Tipe data elemen: int64


*4.2. Mengubah Bentuk Array (.reshape)*

In [9]:
import numpy as np

# Array 1D dengan 12 elemen (0 sampai 11)
arr = np.arange(12)
print(f"Array asli (1D): {arr}")

# Mengubah bentuknya menjadi matriks 3 baris dan 4 kolom
arr_reshaped = arr.reshape(3, 4)
print("Array setelah di-reshape (3, 4):\n", arr_reshaped)

Array asli (1D): [ 0  1  2  3  4  5  6  7  8  9 10 11]
Array setelah di-reshape (3, 4):
 [[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]


**Bab 5: Perhitungan Statistik Dasar**

In [17]:
import numpy as np

data_nilai = np.array([50, 85, 92, 65, 78, 95])
data_jam_belajar = np.array([3, 8, 9, 4, 6, 10])

# Menghitung nilai rata-rata (mean)
print(f"Rata-rata: {np.mean(data_nilai)}")

# Menghitung median (nilai tengah)
print(f"Median dari data nilai: {np.median(data_nilai)}")

#atau bisa menggunakan percentile
print(f"Median(Percentile) dari data nilai: {np.percentile(data_nilai, 50)}")

# Menggabungkan kedua set data menjadi satu array 2D
# Setiap baris adalah pengamatan, setiap kolom adalah variabel
data_gabungan = np.vstack((data_nilai, data_jam_belajar)).T

print("Data Gabungan (Nilai vs Jam Belajar):\n", data_gabungan)

# Menghitung matriks korelasi, untuk mengukur seberapa kuat hubungan linear antara dua set data
print(f"Matriks Korelasi: {np.corrcoef(data_gabungan.T)}") # Transpose diperlukan karena np.corrcoef mengharapkan variabel sebagai baris

# Menghitung matriks kovarians, mengukur bagaimana dua variabel berubah bersama-sama.
#Jika positif, mereka cenderung bergerak ke arah yang sama.
#Jika negatif, mereka bergerak ke arah yang berlawanan.
print(f"Matriks Kovarians: {np.cov(data_jam_belajar, data_nilai)}")

# Menjumlahkan semua nilai
print(f"Jumlah: {np.sum(data_nilai)}")

# Mencari nilai tertinggi
print(f"Nilai Maksimum: {np.max(data_nilai)}")

# Mencari nilai terendah
print(f"Nilai Minimum: {np.min(data_nilai)}")

# Menghitung standar deviasi
print(f"Standar Deviasi: {np.std(data_nilai)}")

# Menghitung varians dari data
print(f"Nilai Varians: {np.var(data_nilai)}")

Rata-rata: 77.5
Median dari data nilai: 81.5
Median(Percentile) dari data nilai: 81.5
Data Gabungan (Nilai vs Jam Belajar):
 [[50  3]
 [85  8]
 [92  9]
 [65  4]
 [78  6]
 [95 10]]
Matriks Korelasi: [[1.       0.976327]
 [0.976327 1.      ]]
Matriks Kovarians: [[  7.86666667  47.2       ]
 [ 47.2        297.1       ]]
Jumlah: 465
Nilai Maksimum: 95
Nilai Minimum: 50
Standar Deviasi: 15.73478100684383
Nilai Varians: 247.58333333333334


**Bab 6: Studi Kasus: Contoh Latihan dengan Jawaban**

*Soal #1: Analisis Data Penjualan*

Buatlah sebuah array NumPy yang merepresentasikan data penjualan harian (dalam ribu Rupiah) sebuah toko selama seminggu: [150, 200, 180, 220, 250, 300, 280]. Hitunglah total penjualan selama seminggu, rata-rata penjualan harian, dan penjualan tertinggi yang terjadi.

In [11]:
import numpy as np

# 1. Membuat array data penjualan
penjualan = np.array([150, 200, 180, 220, 250, 300, 280])

# 2. Menghitung statistik
total = np.sum(penjualan)
rata_rata = np.mean(penjualan)
tertinggi = np.max(penjualan)

print(f"Total Penjualan: Rp{total} ribu")
print(f"Rata-rata Penjualan: Rp{rata_rata:.2f} ribu")
print(f"Penjualan Tertinggi: Rp{tertinggi} ribu")

Total Penjualan: Rp1580 ribu
Rata-rata Penjualan: Rp225.71 ribu
Penjualan Tertinggi: Rp300 ribu


*Soal #2: Konversi Suhu Sensor*

Anda memiliki data suhu dari sebuah sensor dalam Celcius: [25.5, 26.1, -99.0, 27.2, 24.9, -99.0]. Nilai -99.0 adalah data error. Saring data untuk membuang nilai error, lalu konversi semua suhu yang valid ke Fahrenheit. Rumus: F=C
times
frac95+32.

In [12]:
import numpy as np

# 1. Membuat array data suhu
suhu_celcius = np.array([25.5, 26.1, -99.0, 27.2, 24.9, -99.0])

# 2. Menyaring data error menggunakan boolean indexing
suhu_valid = suhu_celcius[suhu_celcius > 0]
print(f"Suhu valid (Celcius): {suhu_valid}")

# 3. Mengonversi suhu valid ke Fahrenheit (vectorization)
suhu_fahrenheit = suhu_valid * (9/5) + 32

print(f"Suhu dalam Fahrenheit: {np.round(suhu_fahrenheit, 2)}")

Suhu valid (Celcius): [25.5 26.1 27.2 24.9]
Suhu dalam Fahrenheit: [77.9  78.98 80.96 76.82]


**Bab 7: Latihan Mandiri**

*Studi Kasus: Diskon Harga*

Anda memiliki daftar harga beberapa barang dalam sebuah array NumPy: [50000, 120000, 75000, 250000, 95000]. Toko sedang mengadakan diskon sebesar 15%. Buatlah array baru yang berisi harga setelah diskon.

Tugas:

- Buat array NumPy untuk daftar harga tersebut.

- Hitung harga setelah diskon 15% untuk semua barang menggunakan operasi vectorization.

- Tampilkan array harga akhir.